x/upgrade/internal/keeper/keeper.go

```diff
@@ -130,3 +139,65 @@ func (k Keeper) ApplyUpgrade(ctx sdk.Context, plan types.Plan) {
 func (k Keeper) IsSkipHeight(height int64) bool {
 	return k.skipUpgradeHeights[height]
 }
+
+// DumpUpgradeInfoToDisk adds plan height to upgrade-info.json
+func (k Keeper) DumpUpgradeInfoToDisk(height int64, name string) error {
+	upgradeInfoFileDir := k.GetHomeDir()
+	upgradeInfoFilePath, err := EnsureUpgradeInfoFileExists(upgradeInfoFileDir)
+	if err != nil {
+		return err
+	}
+
+	var upgradeInfo store.UpgradeInfo
+	upgradeInfo.Height = height
+	upgradeInfo.Name = name
+
+	info, err := json.Marshal(upgradeInfo)
+
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(upgradeInfoFilePath, info, 0644)
+}
+
+// EnsureUpgradeInfoFileExists checks if upgrade-info file is present
+// if not it creates the file
+func EnsureUpgradeInfoFileExists(upgradeInfoFileDir string) (string, error) {
+	if upgradeInfoFileDir != "" {
+		if _, err := os.Stat(upgradeInfoFileDir); os.IsNotExist(err) {
+			err = os.Mkdir(upgradeInfoFileDir, os.ModePerm)
+			if err != nil {
+				return "", err
+			}
+			return filepath.Join(upgradeInfoFileDir, UpgradeInfoFileName), nil
+		}
+	}
+
+	return filepath.Join(upgradeInfoFileDir, UpgradeInfoFileName), nil
+}
+
+// GetHomeDir returns the height at which the given upgrade was executed
+func (k Keeper) GetHomeDir() string {
+	return k.homePath
+}
+
+// ReadUpgradeInfoFromDisk returns the name and height of the upgrade
+// which is written to disk by the old binary when panic'ing
+// if there's an error in reading the info,
+// it assumes that the upgrade info is not available
+func (k Keeper) ReadUpgradeInfoFromDisk() (upgradeName string, upgradeHeight int64) {
+	var upgradeInfo store.UpgradeInfo
+	upgradeInfoPath := filepath.Join(k.homePath, UpgradeInfoFileName)
```

Comment:
I think this should actually be stored under `data/upgrade-info.json`. I think that would be more consistent with how the Cosmos SDK currently works.

Reply:
Its updated now